packages/common/src/major2-validation.ts

```diff
@@ -0,0 +1,686 @@
+import {
+  Concentrations2,
+  IAndCourse2,
+  ICourseRange2,
+  IOrCourse2,
+  IRequiredCourse,
+  IScheduleCourse,
+  IXofManyCourse,
+  Major2,
+  Requirement2,
+  ScheduleCourse,
+  Section,
+  ResultType,
+  Result,
+  Err,
+  Ok,
+} from "./types";
+import { courseToString } from "./course-utils";
+
+/**
+ * general solution: postorder traversal requirements, producing all solutions at each level.
+ * inductive step: combine child solutions to produce solutions for ourselves
+ */
+
+// ------------------------ TYPES ------------------------
+
+/**
+ * A single solution, containing a list of courseToString(c) + # of credits satisfied (needed for XOM)
+ */
+type Solution = {
+  minCredits: number;
+  maxCredits: number;
+  sol: Array<string>;
```

Comment:
Are we using Array instead of Set because we could take some classes more than once? One minor thing I am thinking while writing the test, since the end solutions' order is based on the order of the classes we have taken, I had to write my expected solutions in a certain order. I doubt this would cause any problems when we actually validate if we just care about if it's ok or errors, but just wanna ask and clarify :]

Reply:
ah yes, it's because there could be more than one of the same course. we could use a map, technically (for course -> count) but we went with this because it's simple